SCADA-LTS/Scada-LTS
# Fix #1734 - XSS Vulnerabilities from ScadaBR

## Overview
This PR is fixing vulnerabilities reported in #1384 and #2092:

- [x] XSS in SQL Panel
- [x] Stored XSS in Point Details - user notes
- [x] [MetaDataSource description is not escaped](https://github.com/SCADA-LTS/Scada-LTS/issues/2092)
- [x] Stored XSS - displaying unescaped current username

## Test Sceniario for SQL Panel
### Test nr 1
The first XSS vulnerability lies within the SQL query panel. The output of the SQL queries within this panel is not properly sanitized. In worst case scenarios, this could allow arbitrary JavaScript to execute under the permissions of an administrator. I was able to successfully fire an alert box by performing the below steps.

1. Log in as a low level user.
2. Navigate to the "Users" panel.
3. Append "<script>alert(1);</script>" to the end of the user's email.
4. Log out and log back in with an "admin" account.
5. Navigate to the "SQL" panel.
6. Execute the command "SELECT * FROM users;"
7. A JavaScript alert box will fire.

The output returned from these SQL queries should be properly sanitized.

### Test nr 2
1. Go to SQL Panel
2. Try to execute this queries:
```
<script>alert(1)</script>
\<a onmouseover=alert(document.cookie)\>xxs link\</a\>
<IMG """><SCRIPT>alert("XSS")</SCRIPT>"\>
\";alert('XSS');//
```
3. If user input is properly sanitized, none of above scripts should execute
4. Make sure the normal SQL commands are working properly
5. Execute this SQL query: ``select COUNT(*) from scadalts.pointValues;``
6. This should result a table with COUNT(*)

## Test Scenario for Point Details - user notes
1. Create an active data source as an admin and provide "read" rights to a low level user. Log out of the admin account.
2. Log in as a low level user.
3. From the watch list click "Point Details."
4. Click "Add a Comment" under "User Notes."
5. Enter "<script>alert(1);</script>" into the text field.
6. Log out of the low level user account.
7. Log in to the system as an administrator and click "Point Details."
8. An alert box will fire.

**Expected behavior**
An alert box should **NOT** fire.

## Test Scenario for MetaDataSource
Steps to reproduce the behavior:

1. Create Point of MetaDataSource with following return value: return <script>alert('Hi!')</script>
2. Move to DataSources page.
3. See an Alert with "Hi" message.

**Expected behavior**
Description field in Data Sources page should escape the script tag and the content should not be executed there.

## Test Scenario for displaying session username
1. Go to users tab and change name of the admin account to ``<script>alert("admin XSS")</script>
2. Save changes and go to other tab - an alert box will fire.

**Expected behavior**
An alert box should **NOT** fire - username is escaped.

## Test Scenario for MetaDataSource

1. Import [this](https://github.com/SCADA-LTS/Scada-LTS/files/8140889/configuration.txt) configuration.
**Warning**: at this moment, importing scripts is broken. It has been reported in #1969 

2. Go to: ``Scripting`` tab and create a new script called: "Script1"
3. Change variable name of the ``Datapoints commands`` to ``dpCommands`` and make sure, the checkbox is selected
4. From "Context points", add ``dp_1`` point. Make sure, the variable name is set to ``p1``
5. Go to Script textbox and paste this script:
```
dpCommands.writeDataPoint("DP_451131", p1.value);
```
6. Make sure the first argument is corresponding to XID of the ``dp_2`` point
7. Save this script
8. Now, go to "Graphical View" and select view called ``gv_1``
9. Edit this view - add ``Script1`` to button called "**sc_test**"
10. Save this view and check if it working

**Expected behavior**
``dp_2`` should have a updating random value. After clicking ``sc_test`` button, value of ``dp_2`` should be saved to a value of ``dp_1`` point.
